#934  Transitive dependencies

volodya-lombrozo opened the pull request:
Closes #934 
If the dependency we found contains any transitive dependencies we fail the build.

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,178 @@
+         * @param dependency Dependency
+         * @return True if all filters were passed
+         */
+        private boolean applyAllFilters(final Dependency dependency) {
```
@volodya-lombrozo Can we rename this method? Like apply, match, test or something else. According to this: https://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

volodya-lombrozo replied:
@Graur Fixed. Please, take a look.

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,178 @@
+            try {
+                final List<Dependency> all = new ArrayList<>(0);
+                if (Files.exists(this.file)) {
+                    Logger.info(this, String.format("Dependencies file: %s", this.file));
```
@volodya-lombrozo Can we set this logger level to `debug`? To avoid output noise

volodya-lombrozo replied:
@Graur Fixed. Please, take a look.

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,178 @@
+                }
+                return all;
+            } catch (final IOException ex) {
+                throw new IllegalStateException(ex);
```
@volodya-lombrozo Can we specify this error message according to this: https://www.yegor256.com/2015/12/01/rethrow-exceptions.html

volodya-lombrozo replied:
@Graur Fixed. Please, take a look.

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,178 @@
+         * @checkstyle ParameterNumberCheck (10 lines)
+         */
+        @SuppressWarnings("PMD.UseObjectForClearerAPI")
+        private static Dependency dependency(
```
@volodya-lombrozo Can we use ctor here? According to this: https://www.yegor256.com/2017/11/14/static-factory-methods.html

volodya-lombrozo replied:
@Graur Fixed. Please, take a look. `Dependency` doesn't have a constructor. I've inlined the method.

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/DependenciesPlugin.java:
```diff
@@ -0,0 +1,57 @@
+     * @param origin Dependency which transitive dependencies we are interested of
+     * @return File with dependencies
+     */
+    Path dependenciesFile(Dependency origin);
```
@volodya-lombrozo Same here. Can we make this name simpler?

volodya-lombrozo replied:
@Graur Fixed. Please, take a look.

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,8 +220,36 @@ private Collection<Dependency> deps() throws IOException {
+     * @param dep Dependency
+     * @return True if has transitice dependencies
+     */
+    private boolean hasTransitiveDependencies(final Dependency dep) {
```
@volodya-lombrozo Same here. Can we make this name simpler?

volodya-lombrozo replied:
@Graur Fixed. Please, take a look.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @since 0.28.11
+     */
+    class FilteredDependencies implements Dependencies {
```
@volodya-lombrozo Following adopted convention maybe worth to name it `DcsFiltered`

volodya-lombrozo replied:
@mximp Fixed. Please, take a look.

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @since 0.28.11
+     */
+    class JsonDependencies implements Dependencies {
```
@mximp Fixed. Please, take a look.

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/DependenciesPlugin.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @since 0.28.11
+ */
+interface DependenciesPlugin {
```
@mximp Fixed. Please, take a look. Renamed to `DcsFIle`. wdyt?

mximp replied:
@volodya-lombrozo This is a base interface - it should be named as an entity.
Suffix notation `Dcs***` is used for concrete implementations.

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/DepgraphMavenPlugin.java:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @since 0.28.11
+ */
+final class DepgraphMavenPlugin implements DependenciesPlugin {
```
@mximp Fixed. Please, take a look. Renamed to `DepgraphDcsFile`. wdyt?

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/TransitiveDependencies.java:
```diff
@@ -0,0 +1,132 @@
+ *
+ * @since 0.28.11
+ */
+final class TransitiveDependencies implements Dependencies {
```
@mximp Fixed. Please, take a look.

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/TransitiveDependencies.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @since 0.28.11
+     */
+    private static final class NoRuntimeDependency implements Predicate<Dependency> {
```
@mximp Fixed. Please, take a look.

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/TransitiveDependencies.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @since 0.28.11
+     */
+    private static final class NoSameDependency implements Predicate<Dependency> {
```
@mximp Fixed. Please, take a look.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/DepgraphDcsFile.java:
```diff
@@ -0,0 +1,129 @@
+import org.twdata.maven.mojoexecutor.MojoExecutor;
+
+/**
+ * Plugin for uploading transitive dependencies.
```
@volodya-lombrozo Is this a plugin?

volodya-lombrozo replied:
@mximp Fixed. The description is changed.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/DepgraphDcsFile.java:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @since 0.28.11
+ */
+final class DepgraphDcsFile implements DcsFile {
```
@volodya-lombrozo should be `DfGraph`?

volodya-lombrozo replied:
`Depgraph` it is the name of the maven plugin that we are using for retrieving dependencies file. I think it's better to leave that name in order to emphasise the source.

volodya-lombrozo replied:
Please, take a look https://github.com/ferstl/depgraph-maven-plugin

volodya-lombrozo replied:
@mximp

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsFile.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @since 0.28.11
+ */
+interface DcsFile {
```
@volodya-lombrozo Is this entity actually related to file? What property of its contract tell us something about the file? As I see from its methods it something with dependencies.
Generally it's not clear what this entity represents.

volodya-lombrozo replied:
@mximp Fixed. The class and his method was renamed.

volodya-lombrozo replied:
@mximp 
```java
interface DependenciesFile {
    Path file(Dependency origin);
}
```

The idea is return the file which contains all transient dependencies of the `origin` dependency.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsFile.java:
```diff
@@ -0,0 +1,57 @@
+     * @param origin Dependency which transitive dependencies we are interested of
+     * @return File with dependencies
+     */
+    Path dependencies(Dependency origin);
```
@volodya-lombrozo How do we reference a file here? Based on description of a method we should supply file at some point but the contract provides no means for that.
Also please check [this post](https://www.yegor256.com/2018/08/22/builders-and-manipulators.html) for methods design

volodya-lombrozo replied:
@mximp Fixed. Renamed to `file()`

mximp replied:
@volodya-lombrozo Can you please reply to the comment above: what this class represents?

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsTransitive.java:
```diff
@@ -0,0 +1,135 @@
+ *
+ * @since 0.28.11
+ */
+final class DcsTransitive {
```
@volodya-lombrozo Is it supposed to implement any interface? It has `Dcs` prefix - supposed to be `Dependencies` implementation?

volodya-lombrozo replied:
It was, but we don't need `all` method in that class and we don't use `DscTransitive` ad decorator. I think its better to extend the name

volodya-lombrozo replied:
Renamed to `TransitiveDependencies`

volodya-lombrozo replied:
@mximp Fixed.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,165 @@
+     *
+     * @return List of Maven Dependencies
+     */
+    List<Dependency> toList();
```
@volodya-lombrozo can be just `list()`

volodya-lombrozo replied:
@mximp Fixed. `all` since we are using `Collection` for now

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,165 @@
+ *
+ * @since 0.28.11
+ */
+interface Dependencies {
```
@volodya-lombrozo Maybe it can implement `Iterable`? As far as I understand it is just a collection of items.

volodya-lombrozo replied:
Yes, but I will use `Collections if you don't mind. I need the `stream()` method

volodya-lombrozo replied:
@mximp Fixed.

mximp replied:
@volodya-lombrozo with `Collection` you will have to implement all of its methods... and do it in every class. Do you really need such a heavy contract in this case?

volodya-lombrozo replied:
@mximp I got your idea. No, I don't think we have to implement `Iterable` class - it's an overdesign. We just don't need it. Moreover, I need only `path`.

yegor256 replied:
@volodya-lombrozo I agree that `Iterable` would be a better interface for `DcsDepgraph`. When you need `.size()`, just use `LengthOf(Iterable)` from Cactoos.

mximp commented:
@volodya-lombrozo Please check some comments on the design and naming above.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/DependenciesFile.java:
```diff
@@ -0,0 +1,57 @@
+     * @param origin Dependency which transitive dependencies we are interested of
+     * @return File with dependencies
+     */
+    Path file(Dependency origin);
```
@volodya-lombrozo It is still awkward design here. We have a **File** object which has _file_ attribute. I would expect a **File** to have _path_. 
Also this method accepts particular dependency - does that mean that returned path would contain a file with dependencies for this dependency only?
And `DependenciesFile` represents a file with all possible transitive dependencies?
How would I see it:
```
Path path = new DependenciesFileImpl(origin).path();
```

volodya-lombrozo replied:
@mximp I've implemented the new structure of classes. Please check. Some points:
1) We don't have `TransitiveDependencies` anymore.
2) Also `DependenciesFile` and connected implementations were removed.
3) Predicates still exists, because if we will define them right into `ResolveMojo` (as lambdas) - we won't be able to test them. Also it's impossible to define them: 
3.1)  as `static field`  - we have to provide particular dependency for `NotSame` predicate. 
3.2) as `factory method` - pure design
3.3) Of course, we can introduce one more specific class, like `DcsTransitive`:
```java
public class DcsTransitive implements Dependencies {

    private final Dependencies delegate;

    public DcsTransitive(final Dependencies delegate, final Dependency origin) {
        this.delegate = new DcsFiltered(
            delegate,
            Arrays.asList(
                new DcsFiltered.NotSame(origin),
                new DcsFiltered.NotRuntime(),
                new DcsFiltered.NotTesting()
            )
        );
    }

    @Override
    public Collection<Dependency> all() {
        return delegate.all();
    }
}
```
but, as I understand, you against it, because it's an overdesign.

mximp replied:
@volodya-lombrozo Actually, it's a very good idea... `DcsTransitive` in this form looks fine in this form. Can you add it and use?

volodya-lombrozo replied:
@mximp Added.

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/TransitiveDependencies.java:
```diff
@@ -0,0 +1,135 @@
+ *
+ * @since 0.28.11
+ */
+final class TransitiveDependencies {
```
@mximp Fixed.

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,8 +215,43 @@ private Collection<Dependency> deps() throws IOException {
+                            String.format("%s contains transitive dependencies", dep)
+                        );
+                    }
+                });
```
@volodya-lombrozo parenthesis should be on new line here

volodya-lombrozo replied:
@mximp fixed

volodya-lombrozo commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsTransitive.java:
```diff
@@ -0,0 +1,63 @@
+import org.apache.maven.model.Dependency;
+
+/**
+ * Test case for {@link org.eolang.maven.DcsTransitive}.
```
:) @mximp fixed

mximp commented:
@Graur do you approve?

Graur commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsDepgraph.java:
```diff
@@ -0,0 +1,218 @@
+            );
+            return this.dir.resolve(name);
+        } catch (final MojoExecutionException ex) {
+            throw new IllegalStateException(ex);
```
@volodya-lombrozo Can we specify this error message according to this: https://www.yegor256.com/2015/12/01/rethrow-exceptions.html

volodya-lombrozo replied:
@Graur Done

Graur reviewed (changes requested):
@volodya-lombrozo Good one, just a small comment from my side:

Graur reviewed (approved):
@volodya-lombrozo Looks good to me, thanks!

volodya-lombrozo commented:
@yegor256 can you merge please?

yegor256 commented on eo-maven-plugin/pom.xml:
```diff
@@ -169,13 +169,11 @@ SOFTWARE.
       <groupId>javax.json</groupId>
       <artifactId>javax.json-api</artifactId>
       <version>1.1.4</version>
-      <scope>runtime</scope>
```
@volodya-lombrozo maybe it's better to replace with `provided`? https://howtodoinjava.com/maven/maven-dependency-scopes/

yegor256 commented on eo-maven-plugin/pom.xml:
```diff
@@ -169,13 +169,11 @@ SOFTWARE.
       <groupId>org.glassfish</groupId>
       <artifactId>javax.json</artifactId>
       <version>1.1.4</version>
-      <scope>runtime</scope>
```
@volodya-lombrozo are you sure? Usually `-api` stays in `compile`, while implementation stays in `runtime`.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsDepgraph.java:
```diff
@@ -0,0 +1,224 @@
+        } catch (final MojoExecutionException ex) {
+            throw new IllegalStateException(
+                String.format(
+                    "Dphgraph. Uploading of dependencies file failed for the dependency %s",
```
@volodya-lombrozo uploading, really? I don't think we upload anything here

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsDepgraph.java:
```diff
@@ -0,0 +1,224 @@
+            } catch (final IOException | IllegalStateException ex) {
+                throw new IllegalStateException(
+                    String.format(
+                        "Exception happens during reading the dependencies from json file %s. %s",
```
@volodya-lombrozo it's better to use a long line here. Qulice won't complain

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsFiltered.java:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @since 0.28.11
+ */
+class DcsFiltered implements Dependencies {
```
@volodya-lombrozo if you would use `Iterable`, you wouldn't need this class, but would be able to use `Filtered` from Cactoos: https://www.javadoc.io/doc/org.cactoos/cactoos/latest/org/cactoos/iterable/package-summary.html

volodya-lombrozo replied:
@yegor256 I've used `Filetered` inside `DcsTransitive` and removed `DcsFiltered`

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -95,6 +98,14 @@ public final class ResolveMojo extends SafeMojo {
+     *
+     * @checkstyle MemberNameCheck (7 lines)
+     */
+    @SuppressWarnings({"PMD.ImmutableField", "PMD.AvoidFieldNameMatchingMethodName"})
```
@volodya-lombrozo this is a valid concern of PMD: AvoidFieldNameMatchingMethodName. Don't use similar names for methods and attrs

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/ResolveMojo.java:
```diff
@@ -202,8 +214,40 @@ private Collection<Dependency> deps() throws IOException {
+     */
+    private void checkTransitive(final Collection<Dependency> deps) {
+        if (this.checkTransitive) {
+            deps.forEach(
```
@volodya-lombrozo compare this:

```
list.forEach(
  a -> {
    doit();
  }
)
```

with this:

```
for (list : a) {
  doit();
)
```

which is more compact and has less indentation?

volodya-lombrozo replied:
@yegor256 fixed.

volodya-lombrozo commented:
@yegor256 I've changed the code according with your suggestions. Take a look, please.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsDepgraph.java:
```diff
@@ -0,0 +1,223 @@
+            } catch (final IOException | IllegalStateException ex) {
+                throw new IllegalStateException(
+                    String.format(
+                        "Exception happens during reading the dependencies from json file %s. Probably the file is absent or you have a wrong json format",
```
@volodya-lombrozo for the future, try to enclose `%s` inside texts with single quotes. This will help readers separate your enclosures from the rest of the text:

```
... reading from JSON file '%s`, probably the file...
```

Also, try to use ONE sentence as error message, without a dot inside (and no dot at the end)

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsTransitive.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    public Collection<Dependency> all() {
+        return StreamSupport.stream(
```
@volodya-lombrozo why do you need streams at all? My suggestion, don't use Java streams at all -- their design is pretty anti-OOP in Java. Better use Cactoos primitives. They do the same, but the code will look much more readable and maintainable.

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/DcsTransitive.java:
```diff
@@ -0,0 +1,138 @@
+                new Filtered<>(
+                    new Filtered<>(
+                        this.delegate.all().iterator(),
+                        new NotSame(this.origin)
```
@volodya-lombrozo if you get rid of streams, you will see that here you don't need such special small classes. You can easily inline them right into this big decorator.

volodya-lombrozo replied:
@yegor256 fixed.

volodya-lombrozo replied:
@yegor256 fixed.

yegor256 commented:
@volodya-lombrozo I still don't understand why can't you use `Iterable`? What's the point of making this new interface `Dependencies`?

volodya-lombrozo commented:
> @volodya-lombrozo I still don't understand why can't you use `Iterable`? What's the point of making this new interface `Dependencies`?

@yegor256 I used `Iterable`, removed streams and fixed exception message.  Also I've tried to simplify `DcsTransitive` by removing small predicate classes. wdyt?

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/Dependencies.java:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @since 0.28.11
+ */
+interface Dependencies extends Iterable<Dependency> {
```
@volodya-lombrozo @yegor256 now a question: do we really need this "marker" interface or maybe we can replace all references to `Iterable<Dependency>`?

volodya-lombrozo commented:
@yegor256 reminder

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/27753-1296787466)

rultor commented:
> @rultor merge

@yegor256 Done! FYI, the full log is [here](https://www.rultor.com/t/27753-1296787466) (took me 8min)
